site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,813 @@
+* Status: Implementable
+
+(See definitions in [Kubernetes KEP][kep-status].
+
+[kep-status]: https://github.com/kubernetes/enhancements/blob/master/keps/NNNN-kep-template/kep.yaml#L9
+
+## Goal
+
+Add an idiomatic GRPCRoute for routing gRPC traffic.
+
+## Non-Goals
+
+While certain gRPC implementations support multiple transports and multiple
+interface definition languages (IDLs), this proposal limits itself to
+[HTTP/2](https://developers.google.com/web/fundamentals/performance/http2) as
+the transport and [Protocol Buffers](https://developers.google.com/protocol-buffers)
+as the IDL, which makes up the vast majority of gRPC traffic in the wild.
+
+## Introduction
+
+At the time of writing, the only official Route resource within the Gateway APIs
+is HTTPRoute. It _is_ possible to support other protocols via CRDs and
+controllers taking advantage of this have started to pop up. However, in the
+long run, this leads to a fragmented ecosystem.
+
+gRPC is a [popular RPC framework adopted widely across the industry](https://grpc.io/about/#whos-using-grpc-and-why).
+The protocol is used pervasively within the Kubernetes project itself as the basis for
+many interfaces, including:
+
+- [the CSI](https://github.com/container-storage-interface/spec/blob/5b0d4540158a260cb3347ef1c87ede8600afb9bf/spec.md),
+- [the CRI](https://github.com/kubernetes/cri-api/blob/49fe8b135f4556ea603b1b49470f8365b62f808e/README.md),
+- [the device plugin framework](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/)
+
+Given gRPC's importance in the application-layer networking space and to
+the Kubernetes project in particular, we must ensure that the gRPC control plane
+configuration landscape does not Balkanize.
+
+### Encapsulated Network Protocols
+
+At the time of writing, the only kind of route officially defined by the Gateway
+APIs is `HttpRoute`. This GEP is novel not only in that it introduces a second
+protocol to route, but also in that it introduces the first protocol
+encapsulated in a protocol already supported by the API.
+
+That is, it _is_ theoretically possible to route gRPC traffic using only `HTTPRoute`
+resources, but there are several serious problems with forcing gRPC users to route traffic at
+the level of HTTP. This is why we propose a new resource.
+
+In setting this precendent, we must also introduce a coherent policy for _when_
+to introduce a custom `Route` resource for an encapsulated protocol for which a
+lower layer protocol already exists. We propose the following criteria for such
+an addition.
+
+- Users of the encapsulated protocol would miss out on significant conventional features from their ecosystem if forced to route at a lower layer.
+- Users of the enapsulated protocol would experience a degraded user experience if forced to route at a lower layer.
+- The encapsulated protocol has a significant user base, particularly in the Kubernetes community.
```

Comment:
No change here, just wanted to point out that this is a great set of rules that we probably should pull out into a separate doc somewhere.

Reply:
Thanks. Any guidance on placement?